### ISSUE TITLE
awsvpc network mode incompatibility with control ecs plugin

### ISSUE THREAD

**Issue by zoemccormick** [greymatter-io/terraform-greymatter-ecs#25]
Documenting this here but it may need to be moved to gm-control:

In our setup, we utilize awsvpc networking mode for the ecs tasks:

- https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-networking.html
- https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#network_mode

One reason to use awsvpc networking mode was to allow service discovery that is finding fabric services and creating route53 records automatically so that control can connect to control-api and sidecars can connect to control with XDS_HOST pointing at the route53 dns record without the need for load balancers. Details on this can be found in part 3 of [this comment](https://github.com/greymatter-io/terraform-greymatter-ecs/issues/16#issuecomment-682638636).

This networking is configured [here](https://github.com/greymatter-io/terraform-greymatter-ecs/blob/92797660356baef9b44af0c62eb7eaed90985567/greymatter/modules/sidecar/tasks.tf#L8), and network configurations specified [here](https://github.com/greymatter-io/terraform-greymatter-ecs/blob/92797660356baef9b44af0c62eb7eaed90985567/greymatter/modules/sidecar/services.tf#L11-L13) (based on [terraform documentation](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecs_service#network_configuration)).

The problem is that using awsvpc sets up an ENI for each task and configures `NetworkInterfaces` for the task **instead** of `NetworkBindings`.  For example - heres a describe task on a task container using bridge mode:

```bash
"containers": [
    {
        "containerArn": "arn:aws:ecs:us-east-2:090224759624:container/a0be78cf-d6e3-4741-9783-ff5b92736a6e",
        "taskArn": "arn:aws:ecs:us-east-2:090224759624:task/9265bcf4-304c-4cbf-a4f5-188c107fddf1",
        "name": "control-api",
        "image": "docker.greymatter.io/development/gm-control-api:latest",
        "runtimeId": "b0e32e202d6ca863ea35ed2d3d89dbf6ae2a8ee04c06546a15900f2b0e5e78ae",
        "lastStatus": "RUNNING",
        "networkBindings": [
            {
                "bindIP": "0.0.0.0",
                "containerPort": 5555,
                "hostPort": 32768,
                "protocol": "tcp"
            }
        ],
        "networkInterfaces": [],
        "healthStatus": "UNKNOWN",
        "cpu": "128",
        "memoryReservation": "128"
    }
],
```

and a task using awsvpc:

```bash

"containers": [
    {
        "containerArn": "arn:aws:ecs:us-east-2:090224759624:container/dadb8462-f39a-44bb-b74f-c1c36269990d",
        "taskArn": "arn:aws:ecs:us-east-2:090224759624:task/ba73551b-0b5e-4f88-ba4b-a00459616da6",
        "name": "control-api",
        "image": "docker.greymatter.io/development/gm-control-api:latest",
        "runtimeId": "1a9b0e889570f73d21eb992bf8f0a1d2e557dc2f980fad668365066c94f1c74b",
        "lastStatus": "RUNNING",
        "networkBindings": [],
        "networkInterfaces": [
            {
                "attachmentId": "d8f1b1b2-6b86-46f8-8619-4a799bee99ea",
                "privateIpv4Address": "10.0.0.73"
            }
        ],
        "healthStatus": "UNKNOWN",
        "cpu": "128",
        "memoryReservation": "128"
    }
],
```

This is a problem because the ECS plugin in gm-control is looking for the `NetworkingBindings` of an ecs container in  aws-sdk-go ([this](https://github.com/aws/aws-sdk-go/blob/fd796c520f7417706600d466f3a354e68d017ccb/service/ecs/api.go#L6927)) in [this function](https://github.com/DecipherNow/gm-control/blob/82b6766edee512b05b3a692050c4b58612d071ac/plugins/aws/ecs_helper.go#L480) - which is calling [this](https://github.com/DecipherNow/gm-control/blob/82b6766edee512b05b3a692050c4b58612d071ac/plugins/aws/ecs_helper.go#L325)) to provide the port for the gm cluster instance.  Since [there are no networkbindings](https://github.com/DecipherNow/gm-control/blob/82b6766edee512b05b3a692050c4b58612d071ac/plugins/aws/ecs_helper.go#L339)  I am (expectedly) seeing the following errors in gm-control:

```bash
{
    "level": "error",
    "service": "gm-control",
    "error": "could not bind to port 10808 in container 'arn:aws:ecs:us-east-2:....'",
    "time": "2020-09-22T14:06:59Z"
}
```

```bash
{
    "level": "error",
    "service": "gm-control",
    "clusterName": "gm-cluster",
    "arn": "arn:aws:ecs:us-east-2:service/control-api-sidecar",
    "error": "container port 10808 not exposed on host {\n  ContainerArn: \"arn:aws:ecs:us-east-2:....\",\n  Cpu: \"128\",\n  HealthStatus: \"UNKNOWN\",\n  Image: \"docker.greymatter.io/development/gm-proxy:latest\",\n  LastStatus: \"PENDING\",\n  MemoryReservation: \"128\",\n  Name: \"control-api-sidecar\",\n  NetworkInterfaces: [{\n      AttachmentId: \"2992d0a9-71ee-44e5-9018-c2070352fd2b\",\n      PrivateIpv4Address: \"10.0.0.90\"\n    }],\n  TaskArn: \"arn:aws:ecs:us-east-2:....\"\n} not found",
    "time": "2020-09-22T14:07:29Z"
}
```

**Comment by zoemccormick** [comment]
Potential solutions would be to rework this repo to use bridge mode (this would require a lot of changes) or make updates to the plugin in gm control to check for network interfaces if no bindings are found and get the port a different way [here](https://github.com/DecipherNow/gm-control/blob/82b6766edee512b05b3a692050c4b58612d071ac/plugins/aws/ecs_helper.go#L480) - I would prefer this option because it also would allow an update to the plugin to be able to handle this networking mode in any setup

**Comment by justincely** [comment]
@zoemccormick can you summarize the slack discussion here for posterity?

**Comment by zoemccormick** [comment]
Plans to move forward with changes to the gm-control ecs plugin updates:

Since each task definition can only explicitly have one networking mode, and the networking mode should be able to be determined from the [ecs state](https://github.com/DecipherNow/gm-control/blob/82b6766edee512b05b3a692050c4b58612d071ac/plugins/aws/ecs.go#L145) information, make updates:

1. Have gm-control determine the mode from the state before [this line](https://github.com/DecipherNow/gm-control/blob/82b6766edee512b05b3a692050c4b58612d071ac/plugins/aws/ecs_helper.go#L480)
2. Based on the networking type, determine the hostPort for the instance accordingly.

**Comment by zoemccormick** [comment]
closed with https://github.com/DecipherNow/gm-control/pull/404
